src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -799,17 +799,15 @@ public NodeWarningProperties GetIntersect(NodeWarningProperties other)
                 }
                 else if (thisPackages != null && otherPackages != null)
                 {
-                    packages = new Dictionary<string, HashSet<NuGetLogCode>>(StringComparer.OrdinalIgnoreCase);
+                    var sizeUpperBound = Math.Max(thisPackages.Count, otherPackages.Count);
+                    packages = new Dictionary<string, HashSet<NuGetLogCode>>(sizeUpperBound, StringComparer.OrdinalIgnoreCase);
 
-                    var allKeys = thisPackages.Keys.Concat(otherPackages.Keys)
```

Comment:
The only reason I can see for the old implementation is if `thisPackages` or `otherPackages` can have a different StringComparer to OrdinalIgnoreCase.

Can this ever happen? I can't see tests for it